### PR TITLE
voxel updates

### LINKED ISSUE #1446: [Feature] Voxel switch
Now that voxel support has become a prominent feature of this port, I'm wondering whether it might make sense to add a menu option which allows you to toggle voxels. 

Reasoning:
Like this, users could place the voxel pack in their autoload folders and turn it on or off as they please. Especially useful in case a PWAD replaces the default sprites in some way.

### PR DESCRIPTION
**Pull request by rfomin**
* Don't rotate dropped weapons

* Dynamically increase/decrease maximum distance (improves performance in WADs like nuts.wad)

* Load *.pk3 from *.zip (load VoxelDoom_v2.3.zip directly)

Fix #1446

### CODE REVIEW

File: src/i_video.c [hunk 1]
```diff
@@ -638,6 +639,15 @@ void I_DynamicResolution(void)
         return;
     }
 
+    if (newheight < oldheight)
```
Comment (fabiangreffrath) [inline src/i_video.c:642]: So, we do not directly link maximum distance to rendering resolution. But when it's necessary to decrease resolution, we decrease maximum distance as well and vice versa. This means, we could end up with different maximum distances for the same resolution, depending on which direction we reached the resolution from. Right?

Comment (rfomin) [reply src/i_video.c:642]: > This means, we could end up with different maximum distances for the same resolution, depending on which direction we reached the resolution from. Right?

Right, but that is not a problem, I think. What I have tried to do here is link the max distance and the frame time.

Comment (fabiangreffrath) [reply src/i_video.c:642]: Sounds reasonable, thank you!

**Review by fabiangreffrath — APPROVED** [review]

File: src/r_voxel.c [hunk 2]
```diff
@@ -319,8 +319,25 @@ void VX_Init (void)
 
 #define VX_MINZ         (   4 * FRACUNIT)
 #define VX_MAX_DIST     (2048 * FRACUNIT)
+#define VX_MIN_DIST     ( 512 * FRACUNIT)
```
Comment (rfomin) [inline src/r_voxel.c:322]: @fabiangreffrath Have you tried nuts.wad? The `512` distance is very noticeable here. Maybe we should choose different values?

Comment (fabiangreffrath) [reply src/r_voxel.c:322]: No, not yet. Maybe the steps in maximum distance should be a bit more fine-grained? In the end, by multiplying with 2, we only have three steps: 512, 1024 and 2048.

Comment (rfomin) [reply src/r_voxel.c:322]: > Maybe the steps in maximum distance should be a bit more fine-grained? In the end, by multiplying with 2, we only have three steps: 512, 1024 and 2048.

Maybe. But usually big maps have a lot of monsters, and we will always have minimum distance and minimum resolution there.

Another problem is piles of corpses, lots of models in one place slow down the game a lot. I tried to limit the number of models, but it's too noticeable ("pop-in" like in modern games).

Comment (fabiangreffrath) [reply src/r_voxel.c:322]: We could render corpses as sprites once we reached VX_MIN_DIST.

Comment (rfomin) [reply src/r_voxel.c:322]: > We could render corpses as sprites once we reached VX_MIN_DIST.

Yes, but when voxel corpses are swapped for sprites, it looks jarring. Anyway, I'm looking for feedback, personally I like to sacrifice everything for framerate 😄

Comment (fabiangreffrath) [reply src/r_voxel.c:322]: Now you have the choice of two evils: Noticeable switches from corpse voxels to sprites for an increase in framerate, or continue rendering corpse voxels at the cost of frame rate.

Comment (rfomin) [reply src/r_voxel.c:322]: > Now you have the choice of two evils: Noticeable switches from corpse voxels to sprites for an increase in framerate, or continue rendering corpse voxels at the cost of frame rate.

Exactly, and I feel uncomfortable deciding this myself. I'm one of those people who played Quake 2 almost without textures 😆

File: src/m_menu.c [hunk 3]
```diff
@@ -3539,6 +3547,10 @@ setup_menu_t enem_settings1[] =  // Enemy Settings screen
 
   {"Cosmetic", S_SKIP|S_TITLE, m_null, M_X, M_SPC},
 
+  // [FG] colored blood and gibs
+  {"Voxels", S_YESNO|S_STRICT, m_null, M_X, M_SPC,
+   {"voxels_rendering"}},
```
[the remark this answers is not in the document]
Comment (ceski-1) [reply src/m_menu.c:3552]: > Should we move this to General->Display?

Sure, it makes more sense there to me.

[the remark this answers is not in the document]
Comment (fabiangreffrath) [reply src/m_menu.c:3552]: Do we still have some room in the General menu? 😬

[the remark this answers is not in the document]
Comment (ceski-1) [reply src/m_menu.c:3552]: The big picture question though is if it makes sense to give special treatment for voxels for toggling. That's the case with brightmaps, but there's also the argument that the voxel pack should just be moved out of the autoload folder if you're not using it.

[the remark this answers is not in the document]
Comment (rfomin) [reply src/m_menu.c:3552]: > The big picture question though is if it makes sense to give special treatment for voxels for toggling.

I think yes, it's fun:

https://github.com/fabiangreffrath/woof/assets/5077629/43252662-0a79-476e-8f11-9d5936eb9bd9

File: src/hu_stuff.c [hunk 4]
```diff
@@ -1237,8 +1237,10 @@ static void HU_widget_build_rate (void)
 {
   char hud_ratestr[HU_MAXLINELENGTH];
 
-  sprintf(hud_ratestr, "Sprites %4d Segs %4d Visplanes %4d FPS %3d %dx%d",
-          rendered_vissprites, rendered_segs, rendered_visplanes, fps, video.width, video.height);
+  sprintf(hud_ratestr,
+          "Sprites %4d Segs %4d Visplanes %4d   \x1b%cFPS %3d %dx%d\x1b%c",
+          rendered_vissprites, rendered_segs, rendered_visplanes,
+          '0'+CR_GRAY, fps, video.width, video.height, '0'+CR_ORIG);
```
[the remark this answers is not in the document]
Comment (fabiangreffrath) [reply src/hu_stuff.c:1243]: You could use direct alignment in `y`.

[the remark this answers is not in the document]
Comment (fabiangreffrath) [reply src/hu_stuff.c:1243]: Like this:

```patch
--- a/src/hu_stuff.c
+++ b/src/hu_stuff.c
@@ -154,7 +154,7 @@ static hu_widget_t boom_widgets[MAX_HUDS][MAX_WIDGETS_B] = {
     {&w_sttime, align_left,  align_top},
     {&w_coord,  align_right, align_top},
     {&w_fps,    align_right, align_top},
-    {&w_rate,   align_left,  align_top},
+    {&w_rate,   align_left,  align_direct, 0, 8},
     {NULL}
   }, {
     {&w_armor,  align_left,  align_bottom},

```

[the remark this answers is not in the document]
Comment (rfomin) [reply src/hu_stuff.c:1243]: I tried, but the second line doesn't work for some reason. It doesn't matter.

Is this OK to merge?

[the remark this answers is not in the document]
Comment (fabiangreffrath) [reply src/hu_stuff.c:1243]: > I tried, but the second line doesn't work for some reason. It doesn't matter.

Erg, yeah right, that's a problem.

> Is this OK to merge?

Fine with me!

**Pull request merged**